Implement jps: java process status

pdbain-ibm opened the pull request:
Add Java code to implement the main function.  This uses Attach API to acquire the list of processes
and query their information.

Update Attach API to return JVM arguments

Launchers will be added using coordinated changes in the extension repos.

Requires (mutually):

- https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/267 for Java 8 launcher
- https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/130 for Java 11 launcher
- https://github.com/ibmruntimes/openj9-openjdk-jdk12/pull/12 for Java 12 launcher
- https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/74 for future release launcher

These required test utility changes should be merged before this is merged:
- ~https://github.com/eclipse/openj9/pull/4672 for testing~ Not required
- ~https://github.com/eclipse/openj9/pull/4639 for testing.~ Merged

Doc issues:
- https://github.com/eclipse/openj9-docs/issues/210
- eclipse/openj9-docs#216

Fixes https://github.com/eclipse/openj9/issues/2377

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

pdbain-ibm commented:
To build a launcher, add `closed/make/closed_make/launcher/Launcher-openj9.diagnostics.gmk` in the extensions repository containing the following:
```
include LauncherCommon.gmk

$(eval $(call SetupBuildLauncher, jps, \
    MAIN_MODULE := openj9.diagnostics, \
    MAIN_CLASS := openj9.tools.attach.diagnostics.Jps, \
    CFLAGS := -DEXPAND_CLASSPATH_WILDCARDS, \
    JAVA_ARGS := -Djdk.attach.allowAttachSelf=true, \
))
```

pdbain-ibm commented:
Related issues: 
https://github.com/eclipse/openj9/issues/2259, 
https://github.com/eclipse/openj9/pull/2621, 
https://github.com/eclipse/openj9/issues/2378,
https://github.com/eclipse/openj9/issues/2377.

pshipton commented:
I prefer to have the launchers and tests before this is merged, so they can be run to confirm everything is working properly.

pshipton commented on jcl/src/openj9.diagnostics/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,131 @@
+				+ "-v: print the Java VM arguments, including those produced automatically";
+		for (String a: args) {
+			switch (a) {
+			case "-J": break;
```
I don't think this is required, as any -J argument should be stripped before this code runs.

pdbain-ibm replied:
Fixed.

pshipton commented on jcl/src/openj9.diagnostics/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,131 @@
+			break;
+			case "-h": /* FALLTHROUGH */
+			case "-help": /* FALLTHROUGH */
+			default: System.out.println(HELPTEXT);
```
Does this work with the embedded `\n` characters for all platforms? Some platforms have different line separators.

pdbain-ibm replied:
Good idea.

pshipton commented on jcl/src/openj9.diagnostics/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,131 @@
+					try {
+						VirtualMachine theVm = theProvider.attachVirtualMachine(vmd);
+						try {
+						Properties vmProperties = theVm.getSystemProperties();
```
Indenting looks off here.

pdbain-ibm replied:
Fixed.

pshipton commented:
Does the sun.tools.jps.Jps implementation work on OpenJ9? Is there any need to write another one?

If we do, I'm thinking it should go into the jdk.jcmd module for compatibility, instead of openj9.diagnostics.

pshipton commented:
@keithc-ca

pshipton commented:
I'm thinking any new module we use needs to be added to the j9modules.txt to ensure everything it built in the correct order.

pdbain-ibm commented:
> Does the sun.tools.jps.Jps implementation work on OpenJ9? Is there any need to write another one?

No, the existing `jps` silently produces no results when compiled with OpenJ9.  This utility predates,and does not use, attach API; it appears to use an undocumented proprietary mechanism. It is also not available on Windows.

> If we do, I'm thinking it should go into the jdk.jcmd module for compatibility, instead of openj9.diagnostics.

Will do.

pdbain-ibm commented on jcl/.classpath:
```diff
@@ -37,6 +37,7 @@
 	<classpathentry excluding="**/module-info.java" kind="src" path="src/openj9.gpu/share/classes"/>
 	<classpathentry excluding="**/module-info.java" kind="src" path="src/openj9.sharedclasses/share/classes"/>
 	<classpathentry excluding="**/module-info.java" kind="src" path="src/openj9.traceformat/share/classes"/>
+	<classpathentry excluding="**/module-info.java" kind="src" path="src/openj9.diagnostics/share/classes"/>
```
Done.  There are other items which are out of order which I have not changed.

pdbain-ibm commented on jcl/src/java.base/share/classes/com/ibm/tools/attach/target/Attachment.java:
```diff
@@ -35,6 +36,8 @@
 import java.util.Objects;
 import java.util.Properties;
+
+
```
Fixed.

pdbain-ibm commented on jcl/src/openj9.diagnostics/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,131 @@
+import com.sun.tools.attach.VirtualMachineDescriptor;
+import com.sun.tools.attach.spi.AttachProvider;
+
+/*******************************************************************************
```
Fixed.

pdbain-ibm commented on jcl/src/openj9.diagnostics/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,131 @@
+						if (noPackageName && (parts.length > 0)) {
+							String commandName = parts[0];
+							int finalDot = commandName.lastIndexOf('.');
+							parts[0] = commandName.substring(finalDot+1); /* if the dot is missing, we get the whole string */
```
Fixed.

pdbain-ibm commented on jcl/src/openj9.diagnostics/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,131 @@
+		noPackageName = true;
+		vmidOnly = false;
+		final String HELPTEXT = "jps: Print a list of Java processes and information about them\n"
+				+ "-J: supply arguments to the Java VM running jps\n"
```
Good plan.

pdbain-ibm commented:
> I'm thinking any new module we use needs to be added to the j9modules.txt to ensure everything is built in the correct order.

Done

keithc-ca commented:
I don't see the changes you've claimed to have made - will you be pushing to your branch shortly?

pdbain-ibm commented:
> I don't see the changes you've claimed to have made - will you be pushing to your branch shortly?

I am testing the changes now and will advise when the updates are pushed.

pshipton commented:
Note that Keith advised me the j9modules.txt was only needed for IBM builds and is now obsolete.

pdbain-ibm commented:
Feature code updated per reviewers' comments.

I will advise when the test code has been added to this request.

Launcher changes:
- Java 11: https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/130

pdbain-ibm commented:
Jps implementation changes rae updated and ready for review, as are the launcher definitions.
Test code is in progress.

pshipton commented on jcl/src/jdk.jcmd/share/classes/module-info.java.extra:
```diff
@@ -0,0 +1,23 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+	exports openj9.tools.attach.diagnostics;
```
I don't think this needs to be exported in order to Jps to work. For example, sun.tools.jps is not exported from jdk.cmd

pshipton commented on jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,134 @@
+			break;
+			case "-v": printJvmArguments = true;
+			break;
+			case "-h": /* FALLTHROUGH */
```
Since everything else falls through to print help, it isn't necessary to check for `-h` or `-help`.

pshipton commented:
FYI I note that running `jcmd` with no arguments prints output like running `jps` with no arguments. I suppose it can call `Jps.main(new String[0])`.

pdbain commented:
Updated per reviewers' comments, unit test added, designer tested on MacOS.

keithc-ca commented on test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJps.java:
```diff
@@ -0,0 +1,166 @@
+		assertTrue(SPURIOUS_OUTPUT + searchResult.get(), searchResult.get().equals(vmId));
+		searchResult = search(tgtMgr1.targetId, jpsOutput);
+		assertTrue(CHILD_IS_MISSING, search(tgtMgr1.targetId, jpsOutput).isPresent());
+		assertTrue(SPURIOUS_OUTPUT + searchResult.get(), searchResult.get().equals(tgtMgr1.targetId));
```
Use of `assertEquals` is preferred because when it fails it present both values for comparison.

pdbain commented on jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,133 @@
+								String jvmArguments = vmProperties.getProperty(SUN_JVM_ARGS);
+								if ((null != jvmArguments) && !jvmArguments.isEmpty()) {
+									outputBuffer.add(" "); //$NON-NLS-1$
+									outputBuffer.add(jvmArguments);
```
True about the extra space. If I used `StringBuilder`, I would need to add the spaces manually, while `join()` does that automatically.

keithc-ca replied:
You were already add the spaces (just twice).

pdbain-ibm replied:
> You were already add the spaces (just twice).

Yup.  Fixed that case.  Thanks for catching it.

keithc-ca replied:
I repeat my preference for using a `StringBuilder` rather than a `List<String>` and an array of `String`. As written it is needlessly circuitous.

pdbain commented on jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,133 @@
+		for (String a: args) {
+			switch (a) {
+			case "-l": noPackageName = false;
+			break;
```
Will do.

pdbain commented on test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java:
```diff
@@ -52,6 +59,10 @@ protected void logStackTrace(Exception e) {
 		logger.error(buff.toString());
 	}
 
+	protected void log(String outLine) {
```
Thank you.

pdbain-ibm commented on test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java:
```diff
@@ -155,4 +166,44 @@ protected boolean checkLoadAgentOutput(String errOutput,
+		ProcessBuilder jpsProcess = new ProcessBuilder(cmdLine);
+		Process proc = jpsProcess.start();
+		BufferedReader jpsOutReader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
+		ArrayList<String> outputLines = new ArrayList<>();
```
Nice idea.  Thanks.

pdbain-ibm commented:
Updated per @keithc-ca 's comments.

pshipton commented:
@llxia please review the tests.

llxia commented on test/functional/Java8andUp/playlist.xml:
```diff
@@ -405,6 +405,26 @@
+	$(TEST_STATUS)</command>
+		<levels>
+			<level>sanity</level>
+			<level>extended</level>
```
- Should it be either sanity or extended?
- Please add impls

pdbain-ibm replied:
> * Should it be either sanity or extended?

Both.  I annotated one test as "level.sanity" and the more extensive tests "level.extended".

> * Please add impls

Done

smlambert replied:
Please separate into to distinct targets if you want some in sanity and some in extended.  Highly encourage keeping them together in extended.

pdbain-ibm replied:
Thanks @smlambert .  All tests moved to extended.

keithc-ca commented on jcl/src/java.base/share/classes/com/ibm/tools/attach/target/Attachment.java:
```diff
@@ -22,6 +22,7 @@
  *******************************************************************************/
 package com.ibm.tools.attach.target;
 
+import com.ibm.oti.vm.VM;
```
Only one place benefits from this import; consider removing it or removing the redundant qualification where it is referenced.

keithc-ca commented on jcl/src/java.base/share/classes/com/ibm/tools/attach/target/Attachment.java:
```diff
@@ -35,6 +36,8 @@
 import java.util.Objects;
 import java.util.Properties;
+
+
```
Why these two blank lines?

keithc-ca commented on jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,136 @@
+		int rc = 0;
+		parseArguments(args);
+		AttachProvider theProvider = AttachProvider.providers().get(0);
+		if (null == theProvider) {
```
If there are no providers, I expect the line above threw `IndexOutOfBoundsException`.

keithc-ca replied:
This has not been fixed.

keithc-ca commented on jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,136 @@
+			System.err.println("no attach providers available"); //$NON-NLS-1$
+			rc = 1;
+		} else {
+			ArrayList<String> outputBuffer = new ArrayList<>();
```
Use `List` rather than `ArrayList` for the variable type, like the next line.

keithc-ca commented on jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,136 @@
+							Properties vmProperties = theVm.getSystemProperties();
+							String theCommand = vmProperties.getProperty(SUN_JAVA_COMMAND, ""); //$NON-NLS-1$
+							String parts[] = theCommand.split("\\s+", 2); /* split into at most 2 parts: command and argument string */  //$NON-NLS-1$
+							if (noPackageName && (parts.length > 0)) {
```
`parts.length` will never be zero.

keithc-ca commented on jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/Jps.java:
```diff
@@ -0,0 +1,136 @@
+				break;
+				/* implicitly handle -h and -help via the default case */
+			default: System.out.printf(HELPTEXT);
+			System.exit(1);
```
indentation, missing `break` statement

keithc-ca commented on test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java:
```diff
@@ -155,4 +167,39 @@ protected boolean checkLoadAgentOutput(String errOutput,
+		List<String> outputLines = Collections.emptyList();
+		try (BufferedReader jpsOutReader = new BufferedReader(new InputStreamReader(proc.getInputStream()))) {
+			outputLines = jpsOutReader.lines().collect(Collectors.toList());
+			outputLines.remove(outputLines.size() - 1); /* trim the final null */
```
No null entry should exist.

keithc-ca commented on test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJps.java:
```diff
@@ -0,0 +1,169 @@
+	private static final String CHILD_PROCESS_DID_NOT_LAUNCH = "Child process did not launch"; //$NON-NLS-1$
+	private static final String JPS_COMMAND = "jps"; //$NON-NLS-1$
+	private static final String JPS_Class = "Jps"; //$NON-NLS-1$
+	boolean skipTest;
```
Why is this field not private?

pdbain-ibm commented:
Updated per reviewers' comments and tested on Java 8.

keithc-ca commented on jcl/src/java.base/share/classes/com/ibm/tools/attach/target/Attachment.java:
```diff
@@ -199,7 +200,10 @@ boolean doCommand(InputStream cmdStream, OutputStream respStream) {
-				replyWithProperties(com.ibm.oti.vm.VM.getVMLangAccess().internalGetProperties());
+				Properties internalProperties = com.ibm.oti.vm.VM.getVMLangAccess().internalGetProperties();
+				String argumentString = String.join(" ", com.ibm.oti.vm.VM.getVMArgs()); //$NON-NLS-1$
+				internalProperties.put("sun.jvm.args", argumentString); //$NON-NLS-1$
```
This should create a new `Properites` object rather than modify the system instance.

pdbain-ibm commented:
> This should create a new `Properites` object rather than modify the system instance.

Thanks.  Fixed

> This has not been fixed.

Oops.  Sorry about that.

> I repeat my preference for using a StringBuilder rather than a List<String> and an array of String. As written it is needlessly circuitous.

Okay.

Updated per @keithc-ca 's comments

keithc-ca commented on jcl/src/java.base/share/classes/com/ibm/tools/attach/target/Attachment.java:
```diff
@@ -199,7 +200,11 @@ boolean doCommand(InputStream cmdStream, OutputStream respStream) {
-				replyWithProperties(com.ibm.oti.vm.VM.getVMLangAccess().internalGetProperties());
+				Properties internalProperties = com.ibm.oti.vm.VM.getVMLangAccess().internalGetProperties();
+				String argumentString = String.join(" ", com.ibm.oti.vm.VM.getVMArgs()); //$NON-NLS-1$
+				Properties newProperties = (Properties) internalProperties.clone();
```
Rather than `clone` you should use `new Properties(internalProperties)`.

pdbain-ibm replied:
@keithc-ca I tried that, but `new Properties(internalProperties)` sets `internalProperties` as "defaults", so they don't show up as first-class properties in `Properties.store()`, i.e.
```
jshell> Properties P = System.getProperties()
P ==> {awt.toolkit=sun.lwawt.macosx.LWCToolkit, java.sp ... , java.class.version=55.0}

jshell> Properties p2 = new Properties(P)
p2 ==> {}

jshell> p2.getProperty("java.class.version")
$5 ==> "55.0"

jshell> P.store(System.out,"");
#
#Wed Feb 20 11:33:42 EST 2019
awt.toolkit=sun.lwawt.macosx.LWCToolkit
java.specification.version=11
file.encoding.pkg=sun.io

jshell> p2.store(System.out,"");
#
#Wed Feb 20 11:33:57 EST 2019

```

keithc-ca replied:
That's disappointing.

llxia commented on test/functional/Java8andUp/playlist.xml:
```diff
@@ -405,6 +405,28 @@
+			<group>functional</group>
+		</groups>
+		<impls>
+			<impl>openj9</impl>
```
Does this test apply to ibm implementation? If so, please add `<impl>ibm</impl>`

pdbain-ibm replied:
No, the IBM Java 8 does not support `jps` at present.

llxia replied:
Thanks for clarifying.

pdbain-ibm commented:
Recent changes seem to have broken Java 8.  Investigating...

pshipton commented:
@pdbain-ibm please indicate when this is ready to go, and remove the WIP.

pdbain commented:
Working on it. The first problem was I mucked up `jppd_configuration.xml`.  I fixed that and it passes manual tests, but the unit tests are failing.  Will fix that ASAP, expect late Thursday.

pdbain-ibm commented:
Updated and re-tested on Linux X86  Java 8 & 11.  Changes since last review:
- `jpp_configuration.xml` fixed
- `TestJps.java setupSuite()` updated to create an absolute path to the `jps` executable, searching both `bin` and `jre/bin`
- `playlist.xml` added a Java 8-specific variant of the test with `tools.jar` in the classpath.  Also add an export to the Java 11 variant to suppress error messages.

pshipton commented:
I ran some testing on plinux, which passed, although the jps tests took 20min to run.
I see another change was pushed since this testing to fix the WARNING about illegal reflective access.
```
08:27:34  ...
08:27:34  ... TestNG 6.14.2 by Cédric Beust (cedric@beust.com)
08:27:34  ...
08:27:34  
08:27:34  WARNING: An illegal reflective access operation has occurred
08:27:34  WARNING: Illegal reflective access by org.openj9.test.attachAPI.TargetManager (file:/home/jenkins/jenkins-agent/workspace/Test-Grinder/jvmtest/functional/Java8andUp/GeneralTest.jar) to method com.ibm.tools.attach.target.AttachHandler.waitForAttachApiInitialization()
08:27:34  WARNING: Please consider reporting this to the maintainers of org.openj9.test.attachAPI.TargetManager
08:27:34  WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
08:27:34  WARNING: All illegal access operations will be denied in a future release
08:47:40  PASSED: testApplicationArguments
08:47:40  PASSED: testJpsIdOnly
08:47:40  PASSED: testJpsPackageName
08:47:40  PASSED: testJpsSanity
08:47:40  PASSED: testJvmArguments
```

jdk8: https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/308/
jdk11: https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/306/

pdbain-ibm commented:
> I ran some testing on plinux, which passed, although the jps tests took 20min to run.

The test runs in well under a minute on my Mac and Linux box, and on the internal Linux X86 build farm.  I am testing on a PPCLE machine.  Note that the `AttachAPISanity` exercises much of the same code.  Does it run slowly in a grinder too? Could you try `TestJps` on xLinux?

pdbain-ibm commented:
I ran TestJps on a pLinux box locally; it ran in ~15 seconds.

pshipton commented:
xlinux
jdk8: https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/310/
jdk11: https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/311/

pshipton commented:
Checked AttachAPISanity on plinux from the nightly build last night, but it isn't taking much time to run.
```
02:22:34  ... TestNG 6.14.2 by Cédric Beust (cedric@beust.com)
02:22:34  ...
02:22:34  
02:22:42  PASSED: testAttachToOtherProcess
02:22:42  PASSED: testAttachToSelf
02:22:42  PASSED: testAttachToSelfAndOthers
```

pshipton commented:
~10s on xlinux.

Please request access to the plinux machine ub16-ppcle-1 and check why it is slow to run TestJps.

pshipton commented on test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJps.java:
```diff
@@ -0,0 +1,169 @@
+		commandName = JPS_COMMAND;
+		assertTrue("Attach API failed to launch", TargetManager.waitForAttachApiInitialization()); //$NON-NLS-1$
+		vmId = TargetManager.getVmId();
+		skipTest = !System.getProperty("java.vm.vendor").contains("OpenJ9");  //$NON-NLS-1$//$NON-NLS-2$
```
Is this really required? I'm not sure that it is, since IBM isn't listed as an `<impl>` in the playlist.

pshipton commented:
Windows
jdk8: https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/314/
jdk8 32-bit: https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/313/
jdk11: https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/312/

pdbain-ibm commented:
Ran a personal build on Windows, Java 8 and 11 passed.

pdbain-ibm commented:
> searchResult can be used here.

Right.

> Should there be a tgtMgr2.syncWithTarget()?

Yes.

> Is this really required? I'm not sure that it is, since IBM isn't listed as an `<impl>` in the playlist.

I added it for the case where someone is running testNG without the test tooling.  I removed the flag and changed the check to an assertion so the test fails if you try to run with a different JVM.

pdbain-ibm commented:
Test code updated.

pshipton commented:
Validation test https://ci.eclipse.org/openj9/view/Test/job/Test-Grinder/315/

pshipton commented:
Previous grinder passed.
